### PR TITLE
Fix seq2seq prompt tuning (#439)

### PR DESCRIPTION
**Pull request by glerzing**
The problem is that for prompt tuning, the decoder attention mask is extended as if the prompt was in the decoder (whereas it is in the encoder). So the decoder attention mask was too long. Feel free to propose better ways to fix the problem if my solution is not ideal (I'm not an expert).

See #439 for a code snippet that reproduces the problem.

### CODE REVIEW

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -1091,7 +1091,8 @@ def forward(
             prefix_attention_mask = torch.ones(batch_size, peft_config.num_virtual_tokens).to(
                 decoder_attention_mask.device
             )
-            decoder_attention_mask = torch.cat((prefix_attention_mask, decoder_attention_mask), dim=1)
+            if peft_config.peft_type not in [PeftType.PROMPT_TUNING, PeftType.P_TUNING]:
```
Comment (BenjaminBossan) [inline src/peft/peft_model.py:1094]: Something I wonder about: Should we hard-code the tuning methods that require, or don't require, the concatenation here? Or would it be better to have that knowledge on the model/config itself. Maybe something along the lines of:

```
if peft_config.requires_extending_decoder_mask:
```

Probably, a much better name could be chosen, I'm not sure what is the distinguishing feature between methods that do or don't need this intervention.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for digging into the code and figuring out a solution.

I have a clarifying question about the fix, please take a look at the comment. Also, would it be possible to make a test out of that snippet? We can help if you're not sure about testing.

File: src/peft/peft_model.py (hunk 1 above)
Comment (glerzing) [reply src/peft/peft_model.py:1094]: `PROMPT_TUNING` and `P_TUNING` seem to apply on the encoder prompt, while `PREFIX_TUNING` applies on the decoder's `past_key_values`. Maybe a boolean variable with a name like `self.encoder_tuning` could be considered to replace `in [PeftType.PROMPT_TUNING, PeftType.P_TUNING]`. I personally wouldn't have set it in the peft_config, but rather as an attribute of `PeftModelForSeq2SeqLM`. But I'm also not sure what's the best thing to do here.

**Comment by glerzing** [comment]
> Thanks for digging into the code and figuring out a solution.
> 
> I have a clarifying question about the fix, please take a look at the comment. Also, would it be possible to make a test out of that snippet? We can help if you're not sure about testing.

Makes sense. But actually, I would prefer to avoid spending too much of my time helping Hugging Face. I start to believe that there is an overconfident denial of existential risks there. People are free to have their opinions, but I now personally trust Hugging Face less for figuring out whether its products make a positive impact in the world, which is why I voluntarily help in the first place. Sorry for that, I don't mean to be rude. I still tested manually.

**Review by pacman100 — APPROVED** [review]
Hello @glerzing, Thank you for the fixes, LGTM! 😄

**Review by BenjaminBossan — APPROVED** [review]
To avoid this becoming stale, let's just go ahead and merge it. We can later think about moving the logic elsewhere and adding tests. Thanks @glerzing
